File: src/main/java/org/noxet/noxetserver/menus/inventory/CaptchaSelectionMenu.java
package org.noxet.noxetserver.menus.inventory;

import org.bukkit.entity.Player;
import org.bukkit.event.inventory.ClickType;
import org.bukkit.util.Consumer;
import org.noxet.noxetserver.util.Captcha;
import org.noxet.noxetserver.menus.ItemGenerator;
import org.noxet.noxetserver.util.TextBeautifier;
import org.noxet.noxetserver.util.InventoryCoordinate;
import org.noxet.noxetserver.util.InventoryCoordinateUtil;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class CaptchaSelectionMenu extends InventoryMenu {
    private final List<Captcha.CaptchaSound> captchaSounds;
    private final List<InventoryCoordinate> soundSlots;
    private final Consumer<Captcha.CaptchaSound> callback;

    public CaptchaSelectionMenu(List<Captcha.CaptchaSound> captchaSounds, int soundIndex, int totalSounds, Consumer<Captcha.CaptchaSound> callback) {
        super(3, TextBeautifier.beautify("Sound " + soundIndex + "/" + totalSounds + " - What did you hear?", false), true);

        assert captchaSounds.size() == 3;
        this.captchaSounds = captchaSounds;

        soundSlots = new ArrayList<>();

        for(Captcha.CaptchaSound captchaSound : captchaSounds)
            soundSlots.add(InventoryCoordinateUtil.getCoordinateFromXY(captchaSounds.indexOf(captchaSound) * 2 + 2, 1));

        this.callback = callback;
    }

    @Override
    protected void updateInventory() {
        for(Captcha.CaptchaSound captchaSound : captchaSounds) {
            setSlotItem(
                    ItemGenerator.generateItem(
                            captchaSound.getMaterial(),
                            "§d" + captchaSound.getName(),
                            Collections.singletonList("§e→ Click if you heard this sound.")
                    ), soundSlots.get(captchaSounds.indexOf(captchaSound))
            );
        }
    }

    @Override
    protected boolean onSlotClick(Player player, InventoryCoordinate coordinate, ClickType clickType) {
        for(InventoryCoordinate soundSlot : soundSlots)
            if(coordinate.isAt(soundSlot)) {
                callback.accept(captchaSounds.get(soundSlots.indexOf(soundSlot)));
                return true;
            }

        return false;
    }
}

File: src/main/java/org/noxet/noxetserver/menus/inventory/CaptchaSelectionMenu.scala
package org.noxet.noxetserver.menus.inventory

import org.bukkit.entity.Player
import org.bukkit.event.inventory.ClickType
import org.noxet.noxetserver.util.{Captcha, InventoryCoordinate, InventoryCoordinateUtil, TextBeautifier}
import org.noxet.noxetserver.menus.ItemGenerator

class CaptchaSelectionMenu(question: Captcha.CaptchaQuestion)
  extends InventoryMenu(3,
    TextBeautifier.beautify(s"Sound ${question.soundIndex}/${question.captchaSounds.size} - What did you hear?", false),
    true):
  assert(question.captchaSounds.size == Captcha.answersPerQuestion)

  private val soundSlots: List[(Captcha.CaptchaSound, InventoryCoordinate)] =
    question.captchaSounds.zipWithIndex.map(x => (x._1, InventoryCoordinate(x._2 * 2 + 2, 1)))

  override protected def updateInventory(): Unit =
    soundSlots.foreach(s =>
      setSlotItem(
        ItemGenerator.generateItem(
          s._1.getMaterial,
          s"§d${s._1.getName}",
          List("§e→ Click if you heard this sound.")
        ), s._2))

  override protected def onSlotClick(player: Player, coordinate: InventoryCoordinate, clickType: ClickType): Boolean =
    val clickedSlot = soundSlots.find(_._2 == coordinate)
    clickedSlot.foreach(question.callback(_._1))
    clickedSlot.isDefined